ios/app/brave_core_main.mm

```diff
@@ -106,4 +113,17 @@ - (void)setUserAgent:(NSString*)userAgent {
   _webClient->SetUserAgent(base::SysNSStringToUTF8(userAgent));
 }
 
+- (nullable BraveWalletAPI*)wallet {
+#if BUILDFLAG(BRAVE_WALLET_ENABLED)
+  if (!_wallet) {
+    auto* service =
+        BraveWalletServiceFactory::GetForBrowserState(_mainBrowserState);
+    _wallet = [[BraveWalletAPI alloc] initWithWalletService:service];
```

Comment:
Talked to @bridiver about this, and since this is a class that wraps the service, it should just take the underlying service and not the state, since it doesn't use the state for anything else but getting the service.

History has already been refactored to do the same, and bookmarks will eventually get refactored as well to be equivalent

Reply:
Yeah, but the question is whether or not it needs to be in `BraveCoreMain`. It can still take the `service` but within its own class? @bridiver

Reply:
The idea was to remove the backwards dependency on `GetApplicationContext`. It is similar to using `UIApplication.shared.delegate?.window?.rootViewController` to obtain BVC on iOS. It's not a nice solution, even if it works. `BraveCoreMain` is the owner of the browser state, thus should be the owner of services that are created by using that state (at least linearly). 

We could create an intermediate class such as something like "BraveCoreServices" that houses a browser state and then creates the services there using it, but I'm not sure its needed.

Reply:
The above code your showed is exactly what was refactored out of history as well, and eventually bookmarks

Reply:
Ahh didn't realize we are getting rid of that. I would have preferred BraveCoreMain not doing all of that, similar to `IOSChromeMain`. But I guess it's fine.

Reply:
at least for now BraveCoreMain is the entry point for everything chromium. At some point we probably want to think about how we want to organize things, but it's important that the lifetimes of these services don't extend past the lifetime of BraveCoreMain so this seems reasonable for now

Reply:
Looks to to me then. We can rebase this and merge it if you approve.